<h1 style="text-align: center;">Desafío - Control de Flujo</h1>

- Para realizar este desafío debes haber estudiado previamente todo el material disponible correspondiente a la unidad.
- Una vez terminado el desafío, comprime la carpeta que contiene el desarrollo de los requerimientos solicitados y sube el .zip en el LMS.
- Desarrollo desafío: Individual.
- Puntaje total: 10 puntos.

## Habilidades a evaluar
- Generar arrays ficticios con `numpy`
- Ejecutar un loop que devuelva los resultados requeridos.

## Descripción
¡Has logrado pasar la primera ronda de la selección de la Empresa TI en la cual postulaste en el Desafío anterior! A los reclutadores les interesó tu currículum una vez revisado, pero
siguen existiendo al menos 10 competidores para el puesto, por lo tanto, te solicitan una vez más una serie de requerimientos para poder filtrar técnicamente a los postulantes. Pero esta vez, serán un poco más complejos. Revisémoslos.

## Requerimientos
### 1. Generar un par de arrays ficticios con numpy (1 punto).

**_Tip_**: Utilice la función linspace para generar un array entre 1 y 50, y un array entre 50 y 150.

In [1]:
# Utiliza esta celda para generar los arreglos solicitados
import numpy as np

array1 = np.linspace(1,50) #Por defecto entrega 50 numeros
array2 = np.linspace(51,150,100)


### 2. Ejecuta un loop que devuelva si el número en el primer array es par o impar (1 punto).

_**Tip**_: Utilice la función _módulo_ para encontrar si el número es divisible por 2.

In [2]:
# Utiliza esta celda para programar el loop solicitado
for i in array1:
    if i % 2 == 0:
        print(f'El número {i} es par.')
    else:
        print(f'El nùmero {i} es impar.')

El nùmero 1.0 es impar.
El número 2.0 es par.
El nùmero 3.0 es impar.
El número 4.0 es par.
El nùmero 5.0 es impar.
El número 6.0 es par.
El nùmero 7.0 es impar.
El número 8.0 es par.
El nùmero 9.0 es impar.
El número 10.0 es par.
El nùmero 11.0 es impar.
El número 12.0 es par.
El nùmero 13.0 es impar.
El número 14.0 es par.
El nùmero 15.0 es impar.
El número 16.0 es par.
El nùmero 17.0 es impar.
El número 18.0 es par.
El nùmero 19.0 es impar.
El número 20.0 es par.
El nùmero 21.0 es impar.
El número 22.0 es par.
El nùmero 23.0 es impar.
El número 24.0 es par.
El nùmero 25.0 es impar.
El número 26.0 es par.
El nùmero 27.0 es impar.
El número 28.0 es par.
El nùmero 29.0 es impar.
El número 30.0 es par.
El nùmero 31.0 es impar.
El número 32.0 es par.
El nùmero 33.0 es impar.
El número 34.0 es par.
El nùmero 35.0 es impar.
El número 36.0 es par.
El nùmero 37.0 es impar.
El número 38.0 es par.
El nùmero 39.0 es impar.
El número 40.0 es par.
El nùmero 41.0 es impar.
El número 42.0 es par.
E

### 3. Genere un loop con el segundo array que cuente las siguientes condiciones (1 punto):
- Si el número es divisible por 2 o 3.
- Si el número es divisible por 2 y 3.
- Si el número es divisible por 3 pero no por 2.
- Si el número no es divisible por 2 ni 3.

In [3]:
# Utilza esta celda para programar el loop solicitado, y definir las variables necesarias
for i in array2:
    if i % 3 != 0 and i % 2 != 0: 
        print(f'El número {i} no es divisible ni por 2 ni por 3.')      
    elif i % 2 == 0 and i % 3 == 0:
        print(f'El número {i} es divisivle por 2 y 3.')
    elif i % 3 == 0 and i % 2 != 0:
        print(f'El número {i} es divisible por 3 pero no por 2.')
    else:
        print(f'El número {i} es divisible por 2 o 3.')
    #i % 2 == 0 or i % 3 == 0:


El número 51.0 es divisible por 3 pero no por 2.
El número 52.0 es divisible por 2 o 3.
El número 53.0 no es divisible ni por 2 ni por 3.
El número 54.0 es divisivle por 2 y 3.
El número 55.0 no es divisible ni por 2 ni por 3.
El número 56.0 es divisible por 2 o 3.
El número 57.0 es divisible por 3 pero no por 2.
El número 58.0 es divisible por 2 o 3.
El número 59.0 no es divisible ni por 2 ni por 3.
El número 60.0 es divisivle por 2 y 3.
El número 61.0 no es divisible ni por 2 ni por 3.
El número 62.0 es divisible por 2 o 3.
El número 63.0 es divisible por 3 pero no por 2.
El número 64.0 es divisible por 2 o 3.
El número 65.0 no es divisible ni por 2 ni por 3.
El número 66.0 es divisivle por 2 y 3.
El número 67.0 no es divisible ni por 2 ni por 3.
El número 68.0 es divisible por 2 o 3.
El número 69.0 es divisible por 3 pero no por 2.
El número 70.0 es divisible por 2 o 3.
El número 71.0 no es divisible ni por 2 ni por 3.
El número 72.0 es divisivle por 2 y 3.
El número 73.0 no es divi

### 4. Corrección de errores: Revisa el siguiente código y encuentra los errores a corregir (1 punto):

```python
for i in range(100):
    print(I**2)
```
El error ocurre porque la variable I (mayúscula) no está definida. La variable que sí existe es i (minúscula) que guarda el valor iterado por el for.

<!-- Comenta aquí cómo corregir el error -->

In [4]:
for i in range(100):
    print(i**2)

0
1
4
9
16
25
36
49
64
81
100
121
144
169
196
225
256
289
324
361
400
441
484
529
576
625
676
729
784
841
900
961
1024
1089
1156
1225
1296
1369
1444
1521
1600
1681
1764
1849
1936
2025
2116
2209
2304
2401
2500
2601
2704
2809
2916
3025
3136
3249
3364
3481
3600
3721
3844
3969
4096
4225
4356
4489
4624
4761
4900
5041
5184
5329
5476
5625
5776
5929
6084
6241
6400
6561
6724
6889
7056
7225
7396
7569
7744
7921
8100
8281
8464
8649
8836
9025
9216
9409
9604
9801


### 5. Utilizando la misma base `flights.csv` debes hacer un loop y clasificar los meses con una cantidad de pasajeros menor a la media (3 puntos).  

Para ello:
- Genera un nuevo objeto que represente la media de passengers.

In [5]:
# Utiliza esta celda para guardar la media de passengers
import pandas as pd
import numpy as np

df = pd.read_csv('flights.csv')
media_pasajeros = df['passengers'].mean()

print(media_pasajeros)

280.2986111111111


- Genera una columna en la base de datos que se llame `underperforming` y asígnele `0`.

In [27]:
# Utiliza esta celda para generar la columna solicitada
df['underperforming'] = 0
df.tail()

year      month  passengers  underperforming  outlier
139  1960     August         606                0        1
140  1960  September         508                0        1
141  1960    October         461                0        1
142  1960   November         390                0        0
143  1960   December         432                0        1

- Ejecuta un loop que recorra cada observación de passengers, donde si la observación es menor a la media de passengers, se le asigne a `underperforming` un 1. Para ello utilice la función `at` de pandas

In [7]:
# Utiliza esta celda para programar el loop solicitado
for index, row in df.iterrows():
    if df.at[index,'passengers'] < media_pasajeros:
        df.at[index,'underperforming'] = 1
    else:
        df.at[index,'underperforming'] = 0


### 6. Por último, haz un loop que clasifique los meses donde la cantidad de pasajeros se escapa de la tendencia (3 puntos).

Para ello:
- Genera dos objetos que guarden la media general y la desviación estándar general de `passengers`.

In [8]:
# Utiliza esta celda para generar las variables solicitadas
media_general, desviacion_general = df['passengers'].mean(), df['passengers'].std()
print(media_general, desviacion_general)

280.2986111111111 119.96631694294321


- Genera una nueva columna en la tabla de datos que se llame `outlier` y asígnale 0

In [9]:
# Utiliza esta celda para generar la columna solicitada
df['outlier'] = 0
df.head()

year     month  passengers  underperforming  outlier
0  1949   January         112                1        0
1  1949  February         118                1        0
2  1949     March         132                1        0
3  1949     April         129                1        0
4  1949       May         121                1        0

- Ejecuta un loop que recorra cada observación de passengers, donde si la observación se escapa de la tendencia sea 1, de lo contrario 0.
- Para clasificar los casos que se escapen la tendencia, la observación debe satisfacer _una_ de las siguientes condiciones:
    - La observación debe ser menor a la media menos la desviación estándar.
    - La observación debe ser mayor a la media más la desviación estándar.
- ¿Cuántas observaciones se pueden clasificar como casos extremos?

In [28]:
# Utiliza esta celda para generar el loop solicitado y mostrar los casos extremos
for index, row in df.iterrows():
    if df.at[index,'passengers'] < media_general - desviacion_general or df.at[index,'passengers'] > media_general + desviacion_general:
        df.at[index,'outlier'] = 1
    else:
        df.at[index,'outlier'] = 0

casos = df['outlier'].value_counts()
print(casos)
print(len(df)-1)
print(round(casos[1]/(len(df)-1),4))

0    91
1    53
Name: outlier, dtype: int64
143
0.3706


<!-- Utiliza esta celda para responder con palabras la pregunta indicada -->
Según el resultado del input anterior, del total de casos, 53 son considerados extremos. En otras palabras, el 37,06% son extremos de un total de 143 casos.